### ISSUE TITLE
Fine Tuning NNLM Embedding and Exporting Module Doesn't Add Vocabulary

### ISSUE THREAD

**Issue by meethariprasad** [tensorflow/hub#435]
My assumption was when we fine tune module and export it via estimator, it creates module with updated weights wherever trainable, also adds additional vocabularies to the tokens list for a simple module like nnlm 128 dimension with normalization english. But I found that is not the case when I checked tokens in exported module.

But post fine tune inferences are working as intended (Means the words that I am trying to bring together has been brought together). It also means something changed in assigning vectors through hash buckets when I run OOV words on fine tuned module. Is my understanding right? Can some one please explain what changed for OOV vectors post file tuning?

**Comment by vbardiovskyg** [comment]
Hi @meethariprasad, fine-tuning does change variables as you are mentioning, but does not change the vocabulary. There is currently no API to change the vocabulary of a Hub module or a SavedModel.

You are also noticing correctly that OOV embeddings change. All words that the model does not recognize will be hashed and go to one of the OOV buckets and fine-tuning will be done on the corresponding vectors. This is why, even if the model wouldn't recognize any of the words, it could still in theory learn something meaningful.

**Comment by meethariprasad** [comment]
> Hi @meethariprasad, fine-tuning does change variables as you are mentioning, but does not change the vocabulary. There is currently no API to change the vocabulary of a Hub module or a SavedModel.
> 
> You are also noticing correctly that OOV embeddings change. All words that the model does not recognize will be hashed and go to one of the OOV buckets and fine-tuning will be done on the corresponding vectors. This is why, even if the model wouldn't recognize any of the words, it could still in theory learn something meaningful.

Thanks @vbardiovskyg . This is a great explanation and you have been great help in many answers.

Apologies for reconfirming my understanding of your statement with a simple example below. 

On OOV Part there is some clarity required though.
OOV="abrakadabra"
Step 1: 
1. Embedding without fine tuning: OOV -> Module()-> One of Hash Buckets
Step 2:
2. Export Module post fine Tuning along with OOV word. 

3. Infer Embedding from fine tuned exported module: OOV -> Module()-> Still Same Hash Bucket but now embeddings changed.

Does that mean fine tuning changed the variables associated with hash bucket. (i.e: Fine tuning made sure the hasbucket output vector is fine tuned. Means now onwards any word that comes out of that specific hasbucket will have changed embedding, corresponding to fine tuned process outcome?

**Comment by vbardiovskyg** [comment]
> Does that mean fine tuning changed the variables associated with hash bucket.
Yes, that's right.

> Means now onwards any word that comes out of that specific hasbucket will have changed embedding, corresponding to fine tuned process outcome?
Yes, the change is permanent and applies to all word that hash to that OOV bucket.

**Comment by meethariprasad** [comment]
@

> > Does that mean fine tuning changed the variables associated with hash bucket.
> 
> Yes, that's right.
> 
> > Means now onwards any word that comes out of that specific hasbucket will have changed embedding, corresponding to fine tuned process outcome?
> 
> Yes, the change is permanent and applies to all words that hash to that OOV bucket.

@vbardiovskyg  Thanks a lot again for quick reply.
